File: src/multi/algo/CCGwithPAP_Reactive.java
package multi.algo;

import ilog.concert.IloException;
import ilog.cplex.IloCplex;
import multi.data.InputData;
import multi.data.Parameter;
import multi.model.*;
import multi.data.Scenario;
import java.io.IOException;
import java.util.List;

public class CCGwithPAP_Reactive  extends CCG {
    private DualSubProblemReactive dsp;
    public CCGwithPAP_Reactive(InputData in, Parameter p) throws IloException, IOException {
        super();
        this.in = in;
        this.p = p;
        this.tau = p.getTau();
        this.Algo = "CCG&PAP-Reactive";
        this.AlgoID = Algo + "-R"+ in.getShipRouteSet().size() + "-T" + p.getTimeHorizon() + "-"+ FleetType + "-S" + randomSeed + "-V" + vesselCapacityRange;
        frame();
    }
    public CCGwithPAP_Reactive(InputData in, Parameter p, int tau) throws IloException, IOException {
        super();
        this.in = in;
        this.p = p;
        this.tau = tau;
        this.Algo = "CCG&PAP-Reactive";
        this.AlgoID = Algo + "-R"+ in.getShipRouteSet().size() + "-T" + p.getTimeHorizon() + "-"+ FleetType + "-S" + randomSeed + "-V" + vesselCapacityRange;
        frame();
    }
    @Override
    protected double initialModel() throws IloException, IOException {
        double start = System.currentTimeMillis();

        dsp =new DualSubProblemReactive(in, p, tau);
        mp=new MasterProblem(in, p, "Reactive");

        if(WhetherAddInitializeSce){
            mp.addScene(sce.get(0));
        }

        if(WhetherSetInitialSolution){
            DetermineModel dm = new DetermineModel(in, p);
            mp.setInitialSolution(dm.getVVarValue());
        }

        return System.currentTimeMillis() - start;
    }

    @Override
    protected void frame() throws IloException, IOException {
        initialize();

        if(WhetherCalculateMeanPerformance && UseHistorySolution){
            calculateMeanPerformance();
            return;
        }

        double[] maxDemandVar = p.getMaximumDemandVariation();
        // change  MaxVarDemand
        // beta = min(k, m/k)
        double beta=(double)tau > p.getDemand().length/(double)tau ?
                p.getDemand().length/(double)tau : (double)tau;
        p.changeMaximunDemandVariation(beta);

        if(WhetherAddInitializeSce)
            initializeSce(sce);

        double masterObjective;

        double time0 = System.currentTimeMillis();
        initialModel();

        printIterTitle(fileWriter, System.currentTimeMillis() - time0);
        printIteration(fileWriter, lower[0], upper[0],
                0, 0, 0,
                "--", 0, "--", 0);

        /*initializeSce(sce);*/
        int flag=0;
        double start0 = System.currentTimeMillis();
        while(upperBound - lowerBound > boundGapLimit
                && flag==0
                && iteration<maxIterationNum
                && (System.currentTimeMillis() - start0)/1000 < maxIterationTime
        ){
            // build and solve master model
            // add new scene to Master Problem
            if (iteration != 0)
            {
                mp.addScene(sce.get(sce.size()-1));
            }

            double start1 = System.currentTimeMillis();
            mp.solveReactiveModel();
            double end1 = System.currentTimeMillis();

            // get the solution
            masterObjective = mp.getObjVal();
            this.setOperationCost(mp.getOperationCost());
            this.setTotalCost(mp.getObjVal());

            // check if the mp-solution changed
            if(! addSolutionPool(mp.getSolution())){
                System.out.println("MP solution duplicate");
                flag=1;
            }

            // update lower bound: LB = max {LB, Obj*}
            if(mp.getObjVal()>lowerBound
                    && mp.getSolveStatus() == IloCplex.Status.Optimal) {
                setLowerBound(mp.getObjVal());
            }

            dsp.changeObjectiveVCoefficients(mp.getVVarValue(), mp.getVVarValue2());
            double start2 = System.currentTimeMillis();
            dsp.solveModel();
            double end2 = System.currentTimeMillis();

            if(addScenarioPool(dsp.getScene())){
                sce.add(dsp.getScene());
            }
            else{
                flag=1;
            }

            SubProblemReactive sp = new SubProblemReactive(in, p);
            sp.changeConstraintCoefficients(mp.getVVarValue(), mp.getVVarValue2(), dsp.getScene().getRequest());
            sp.solveModel();
            sp.end();
//            System.out.println("SP-Obj = "+sp.getObjective());

            if(dsp.getObjVal()+mp.getOperationCost() < upperBound
                    && dsp.getSolveStatus() == IloCplex.Status.Optimal)
            {
                setUpperBound(dsp.getObjVal()+mp.getOperationCost());
            }

            iteration++;
            upper[iteration]=upperBound;
            lower[iteration]=lowerBound;
            masterObj[iteration] = masterObjective;
            subObj[iteration] = dsp.getObjVal();

            printIteration(fileWriter, lower[iteration], upper[iteration],
                    end2 - start2, end1 - start1, System.currentTimeMillis() - start0,
                    dsp.getSolveStatusString(), dsp.getMipGap(), mp.getSolveStatusString(), mp.getObjGap());
        }

        setTotalCost(upperBound);
        setOperationCost(mp.getObjVal() - mp.getEtaValue());
        setvValue2(mp.getVVarValue2());

        setAlgoResult();

        end();

        p.setMaximumDemandVariation(maxDemandVar);
    }

    @Override
    protected void initializeSce(List<Scenario> sce)
    {
        double[] sss  =new double [in.getRequestSet().size()];

        // |v| = tau/I |e| = tau/I
        double v=(double)tau/(double)p.getDemand().length * 1/ (Math.sqrt(p.getDemand().length));
        for(int i=0;i<p.getDemand().length;i++)
        {
            sss[i]=v ;
        }

        sce.add(new Scenario(sss));
    }
    public void setvValue2(int[][] vValue2) {
        this.vValue2 = vValue2;
    }
    private int[][] vValue2;

    public void printSolution(){
        System.out.println("Master Objective ="+String.format("%.2f", getObjVal()));

        System.out.print("Vessel Decision vVar (MP) : ");
        for(int r = 0; r<p.getShippingRouteSet().length; r++)
        {
            System.out.print(p.getShippingRouteSet()[r]+"(");
            for(int h=0;h<p.getVesselSet().length;h++)
            {
                if(vValue[h][r] != 0)
                {
                    System.out.print(p.getVesselSet()[h]+")\t");
                }
            }
        }
        System.out.println();
        System.out.print("Reactive Decision vVar2 (MP) : ");
        for(int w=0;w<p.getVesselPathSet().length;w++)
        {
            for(int h=0;h<p.getVesselSet().length;h++)
            {
                if(vValue2[h][w] != 0)
                {
                    System.out.print(p.getVesselPathSet()[w]+"("+p.getVesselSet()[h]+")\t");
                }
            }
        }
        System.out.println();
    }
}

File: Multi_routes_deploy/sources/algo/CCGwithPAP_Reactive.cpp
#include "CCGwithPAP_Reactive.h"

namespace fleetdeployment {

    void CCGwithPAP_Reactive::frame() {
        if (WhetherPrintProcess || WhetherPrintIteration) {
            std::cout << "=========C&CG with PAP (Reactive Strategy) : ===========" << std::endl;
        }

        std::string filename = "CCG-PAP-Reactive" + std::to_string(p_.GetTimePointSet().size()) + ".txt";
        std::ofstream fileWriter(filename, std::ios::app);

        if (!fileWriter.is_open()) {
            std::cerr << "Error: Unable to open file " << filename << std::endl;
        }
        time_t start = time(0);

        std::vector<double> maxDemandVar = p_.GetMaximumDemandVariation();
        // change  MaxVarDemand
        // beta = min(k, m/k)
        double beta = (double)p_.GetDemand().size() / (double)tau;
        if (beta > tau)
        {
            beta = tau;
        }
        p_.changeMaximunDemandVariation(beta);

        SetUpperBound(1e12);
        SetLowerBound(-1e12);

        // add initial scene to make model feasible
        std::vector<Scenario> sce;

        // vvv ¡ª¡ª the solution of master problem
        std::vector<std::vector<int>> vvv(p_.GetVesselSet().size(), std::vector<int>(p_.GetVesselRouteSet().size()));
        double masterObjective;
        int flag = 0;
        int max_iteration = 100;
        std::vector<double> upper(max_iteration + 1);
        std::vector<double> lower(max_iteration + 1);
        std::vector<double> masterObj(max_iteration + 1);
        std::vector<double> subObj(max_iteration + 1);
        upper[0] = upperBound;
        lower[0] = lowerBound;
        int iteration = 0;
        double gap = std::numeric_limits<double>::max();

        time_t time0 = time(0);
        MasterProblem mp(in_, p_, true);
        DualSubProblemReactive dsp(in_, p_, 1);
        Initialize(sce);

        if (WhetherPrintProcess) {
            std::cout << "=========C&CG with PAP-Reactive: ===========" << std::endl;
            std::cout << "BuildModelTime = "
                << std::fixed << std::setprecision(2) << time(0) - time0 << std::endl;
            std::cout << "k" << "\t\t"
                << "UB" << "\t\t"
                << "LB" << "\t\t"
                << "DSP-SolveTime" << "\t\t"
                << "MP-SolveTime" << "\t\t"
                << "Total Time" << "\t\t"
                << "DSP-Status" << "\t\t"
                << "MP-Status" << std::endl;
            std::cout << iteration << "\t\t"
                << std::fixed << std::setprecision(2) << upper[iteration] << "\t\t"
                << std::fixed << std::setprecision(2) << lower[iteration] << std::endl;
        }

        if (WhetherOutputLog) {
            fileWriter << "k" << "\t\t"
                << "UB" << "\t\t"
                << "LB" << "\t\t"
                << "DSP-SolveTime" << "\t\t"
                << "MP-SolveTime" << "\t\t"
                << "Total Time" << "\t\t"
                << "DSP-Status" << "\t\t"
                << "MP-Status" << std::endl;
            fileWriter << iteration << "\t\t"
                << std::fixed << std::setprecision(2) << upper[iteration] << "\t\t\t"
                << std::fixed << std::setprecision(2) << lower[iteration] << std::endl;
        }

        while (upperBound - lowerBound > MIPGapLimit
            && gap > MIPGapLimit
            && flag == 0
            && iteration < max_iteration)
        {
            // build and solve master model
            // add new scene to Master Problem
            mp.addReactiveScene(sce[iteration]);

            time_t start1 = time(0);
            mp.SolveReactiveModel();
            time_t end1 = time(0);

            // Get the solution
            masterObjective = mp.GetObjVal();
            this->SetOperationCost(mp.GetOperationCost());
            this->SetTotalCost(mp.GetObjVal());
            // check if the mp-solution changed
            double temp = 0;
            for (int r = 0; r < p_.GetVesselRouteSet().size(); r++)
            {
                for (int h = 0; h < p_.GetVesselSet().size(); h++)
                {
                    temp = temp + abs(mp.GetVVarValue()[h][r] - vvv[h][r]);
                    vvv[h][r] = mp.GetVVarValue()[h][r];
                }
            }
            // master solution has no change
            if (temp < 0.5)
            {
                flag = 1;
            }

            // update lower bound: LB = max {LB, Obj*}
            if (masterObjective > lowerBound) {
                SetLowerBound(masterObjective);
            }

            dsp.changeObjectiveVCoefficients(mp.GetVVarValue(), mp.GetVVarValue2());
            time_t start2 = time(0);
            dsp.SolveModel();
            time_t end2 = time(0);

            sce.push_back(dsp.GetScene());

            SubProblemReactive sp(in_, p_);
            sp.changeConstraintCoefficients(mp.GetVVarValue(), mp.GetVVarValue2(), dsp.GetScene().GetRequest());
            sp.solveModel();
            sp.end();
            // std::cout<<"SP-Obj = "+sp.GetObjective());

            if (sp.GetObjVal() + mp.GetOperationCost() < upperBound)
            {
                SetUpperBound(sp.GetObjVal() + mp.GetOperationCost());
            }

            iteration = iteration + 1;
            upper[iteration] = upperBound;
            lower[iteration] = lowerBound;
            masterObj[iteration] = masterObjective;
            subObj[iteration] = dsp.GetObjVal();
            gap = upper[iteration] - lower[iteration];

            if (WhetherPrintProcess || WhetherPrintIteration) {
                std::cout << iteration << "\t\t"
                    << std::fixed << std::setprecision(2) << upper[iteration] << "\t\t"
                    << std::fixed << std::setprecision(2) << lower[iteration] << "\t\t"
                    << std::fixed << std::setprecision(2) << difftime(end2, start2) << "\t\t"
                    << std::fixed << std::setprecision(2) << difftime(end1, start1) << "\t\t"
                    << dsp.GetSolveStatusString() << "(" << std::fixed << std::setprecision(4) << dsp.GetMipGap() << ")\t\t"
                    << mp.GetSolveStatusString() << "(" << std::fixed << std::setprecision(4) << mp.GetMipGap() << ")\n";
            }
            if (WhetherOutputLog) {
                fileWriter << iteration << "\t\t"
                    << std::fixed << std::setprecision(2) << upper[iteration] << "\t\t"
                    << std::fixed << std::setprecision(2) << lower[iteration] << "\t\t"
                    << std::fixed << std::setprecision(2) << difftime(end2, start2) << "\t\t"
                    << std::fixed << std::setprecision(2) << difftime(end1, start1) << "\t\t"
                    << dsp.GetSolveStatusString() << "(" << std::fixed << std::setprecision(4) << dsp.GetMipGap() << ")\t\t"
                    << mp.GetSolveStatusString() << "(" << std::fixed << std::setprecision(4) << mp.GetMipGap() << ")\n";
            }

        }

        SetObj(upperBound);
        SetTotalCost(upperBound);
        SetOperationCost(mp.GetObjVal() - mp.GetEtaValue());
        SetIter(iteration);
        SetVValue(mp.GetVVarValue());
        SetVValue2(mp.GetVVarValue2());

        Gap = (upperBound - lowerBound) / lowerBound;

        mp.end();
        dsp.end();

        time_t end = time(0);
        SetSolveTime(difftime(end, start));

        if (WhetherPrintProcess || WhetherPrintIteration) {
            std::cout << "CCG&PAP(Reactive Strategy) SolveTime = " << difftime(end, start) << std::endl;
            std::cout << "==================================" << std::endl;
        }

        if (WhetherOutputLog) {
            fileWriter << "CCG&PAP(Reactive Strategy) SolveTime = " << difftime(end, start) << std::endl;
        }
        p_.SetMaximumDemandVariation(maxDemandVar);
    }

    void CCGwithPAP_Reactive::Initialize(std::vector<Scenario>& sce)
    {
        std::vector<double> sss(in_.GetRequests().size());

        double b1 = (double)tau / (double)(p_.GetDemand().size());

        // beta = min{k , I/k}
        double beta = (double)p_.GetDemand().size() / (double)tau;
        if (beta > tau)
        {
            beta = tau;
        }

        for (int i = 0; i < p_.GetDemand().size(); i++)
        {
            sss[i] = b1 / beta;
        }

        Scenario sp;
        sp.SetRequest(sss);
        sce.push_back(sp);
    }

    void CCGwithPAP_Reactive::PrintSolution() {
        std::cout << "Master Objective =" << std::fixed << std::setprecision(2) << obj << std::endl;

        std::cout << "Vessel Decision vVar (MP) : " << std::endl;
        for (int r = 0; r < p_.GetVesselRouteSet().size(); r++)
        {
            std::cout << p_.GetVesselRouteSet()[r] << ":\t";
            for (int h = 0; h < p_.GetVesselSet().size(); h++)
            {
                if (vValue[h][r] != 0)
                {
                    std::cout << "(" << p_.GetVesselSet()[h] << ")\t";
                }
            }
        }
        std::cout << std::endl;

        std::cout << "Reactive Decision vVar2 (MP) : " << std::endl;
        for (int r = 0; r < p_.GetVesselRouteSet().size(); r++)
        {
            std::cout << p_.GetVesselRouteSet()[r] << ":\t";
            for (int h = 0; h < p_.GetVesselSet().size(); h++)
            {
                if (vValue2[h][r] != 0)
                {
                    std::cout << "(" << p_.GetVesselSet()[h] << ")\t";
                }
            }
        }
        std::cout << std::endl;
    }



}